osgp/platform/osgp-secret-management/src/main/java/org/opensmartgridplatform/secretmanagement/application/services/SecretManagementService.java

```diff
@@ -48,58 +53,51 @@ public SecretManagementService(
-                .forEach(this.secretRepository::save);
-        //@formatter:on
-    }
-
     private DbEncryptionKeyReference getKey() {
         final Date now = new Date();
-        final Page<DbEncryptionKeyReference> keyRefsPage = this.keyRepository.findByTypeAndValid(now,
-                this.encryptionProviderType, Pageable.unpaged());
+        final Page<DbEncryptionKeyReference> keyRefsPage = this.keyRepository.findByTypeAndValid(
+                this.encryptionProviderType, now, Pageable.unpaged());
         if (keyRefsPage.getSize() > 1) {
             throw new IllegalStateException("Multiple encryption keys found that are valid at " + now);
         }
         return keyRefsPage.stream().findFirst().orElseThrow(
                 () -> new NoSuchElementException("No encryption key found that is valid at " + now));
     }
 
-    private TypedSecret validateSecret(final String deviceIdentification, final TypedSecret secret) {
+    private TypedSecret validateNewSecret(final String deviceIdentification, final TypedSecret secret) {
+        this.checkNrNewSecretsOfType(deviceIdentification, secret.getSecretType(), 0);
         if (secret.getSecret() == null) {
             throw new IllegalArgumentException("No secret string set");
         } else if (secret.getSecretType() == null) {
             throw new IllegalArgumentException("No secret type set");
-        } else if (this.isIdenticalToCurrent(deviceIdentification, secret)) {
+        } else if (this.isIdenticalToActive(deviceIdentification, secret)) {
             throw new IllegalArgumentException(
                     String.format("Secret is identical to current secret (%s, %s)", deviceIdentification,
                             secret.getSecretType().name()));
         }
         return secret;
     }
 
-    private boolean isIdenticalToCurrent(final String deviceIdentification, final TypedSecret secret) {
-        final Optional<TypedSecret> current = this.retrieveSecret(deviceIdentification, secret.getSecretType());
-        return current.isPresent() && current.get().getSecret().equals(secret.getSecret());
+    private boolean isIdenticalToActive(final String deviceIdentification, final TypedSecret secret) {
+        final Optional<TypedSecret> active = this.retrieveActiveSecret(deviceIdentification, secret.getSecretType());
+        return active.isPresent() && active.get().getSecret().equals(secret.getSecret());
     }
 
     private DbEncryptedSecret createEncrypted(final String deviceIdentification, final TypedSecret typedSecret,
-                                              final DbEncryptionKeyReference keyReference) {
+            final DbEncryptionKeyReference keyReference) {
         final String secretString = typedSecret.getSecret();
         final byte[] secretBytes = HexUtils.fromHexString(secretString);
-        final Secret secret = new Secret(secretBytes);
+        final Date now = new Date(); //TODO check creation & time zone
```

Comment:
![INFO](https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO') Complete the task associated to this TODO comment. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](http://ci.opensmartgridplatform.org/sonarqube/coding_rules#rule_key=squid%3AS1135)